keep/api/core/dependencies.py

```diff
@@ -206,91 +241,151 @@ def verify_bearer_token_single_tenant(token: str = Depends(oauth2_scheme)) -> st
     # if we don't want to use authentication, return the single tenant id
     if (
         os.environ.get("AUTH_TYPE", AuthenticationType.NO_AUTH.value)
         == AuthenticationType.NO_AUTH.value
     ):
-        return SINGLE_TENANT_UUID
+        return AuthenticatedEntity(
+            tenant_id=SINGLE_TENANT_UUID, email=SINGLE_TENANT_EMAIL
+        )
 
     api_key = __extract_api_key(request, api_key, authorization)
 
     tenant_api_key = get_api_key(api_key)
     if not tenant_api_key:
         raise HTTPException(status_code=401, detail="Invalid API Key")
 
+    role = get_role_by_role_name(tenant_api_key.role)
+    # validate scopes
+    if not role.has_scopes(scopes):
+        raise HTTPException(
+            status_code=403,
+            detail=f"You don't have the required scopes to access this resource [required scoopes: {scopes}]",
+        )
     request.state.tenant_id = tenant_api_key.tenant_id
-    return tenant_api_key.tenant_id
-
-
-def verify_token_or_key(
-    request: Request,
-    api_key: Optional[str] = Security(auth_header),
-    authorization: Optional[HTTPAuthorizationCredentials] = Security(http_basic),
-    token: Optional[str] = Depends(oauth2_scheme),
-) -> str:
-    # Attempt to verify API Key first
-    if api_key:
-        try:
-            return verify_api_key(request, api_key, authorization)
-        except Exception:
-            logger.exception("Failed to validate API Key")
-            raise HTTPException(
-                status_code=401, detail="Invalid authentication credentials"
-            )
-    # If API Key is not present or not valid, attempt to verify the token
-    if token:
-        try:
-            return verify_bearer_token(token)
-        except Exception:
-            logger.exception("Failed to validate token")
-            raise HTTPException(
-                status_code=401, detail="Invalid authentication credentials"
-            )
-    raise HTTPException(status_code=401, detail="Missing authentication credentials")
+    return AuthenticatedEntity(tenant_api_key.tenant_id, tenant_api_key.created_by)
+
+
+def AuthVerifier(scopes: list[str] = []):
+    # Determine the authentication type from the environment variable
+    auth_type = os.environ.get("AUTH_TYPE", AuthenticationType.NO_AUTH.value)
+
+    # Return the appropriate verifier based on the auth type
+    if auth_type == AuthenticationType.MULTI_TENANT.value:
+        return AuthVerifierMultiTenant(scopes)
+    else:
+        return AuthVerifierSingleTenant(scopes)
+
+
+class AuthVerifierMultiTenant:
+    """Handles authentication and authorization for multi tenant mode"""
+
+    def __init__(self, scopes: list[str] = []) -> None:
+        self.scopes = scopes
+
+    def __call__(
+        self,
+        request: Request,
+        api_key: Optional[str] = Security(auth_header),
+        authorization: Optional[HTTPAuthorizationCredentials] = Security(http_basic),
+        token: Optional[str] = Depends(oauth2_scheme),
```

Comment:
If it's not api_key and no token what will happen here? the function will just return None, is that ok?

Reply:
no, its 
if api_key: ..
if token: ..
raise HTTPException(401)

Reply:
<img width="781" alt="CleanShot 2024-01-04 at 14 39 56@2x" src="https://github.com/keephq/keep/assets/12069200/f79b8926-cf25-4906-8d2c-d72d498dabcf">